File: frontend/lib/doc-chat.ts
import { authorizedPost } from "@/lib/api";
import { type ChatMessage } from "@/lib/chat";
import { todayIso } from "@/lib/nda";

export interface DocChatResponseBody {
  reply: string;
  selectedDocument?: string;
  updates: Record<string, string>;
}

export const GREETING =
  "Hi! I can help you draft a legal agreement. What kind of document do you " +
  "need — or tell me about your situation and I'll suggest one.";

/** Fixed opener (no LLM call on page load); still sent as the first
 * transcript message so the model has full context. */
export const initialMessages = (): ChatMessage[] => [
  { role: "assistant", content: GREETING },
];

/** One stateless turn of the generic document chat. `documentKey` null means
 * the conversation is still choosing a document. */
export const postDocChat = (
  transcript: ChatMessage[],
  documentKey: string | null,
  fields: Record<string, string>,
) =>
  authorizedPost<DocChatResponseBody>("/api/doc-chat", {
    transcript,
    documentKey,
    fields,
    today: todayIso(),
  });

File: backend/app/schemas/doc_chat.py
"""Wire contract for POST /api/doc-chat (the generic document chat).

`documentKey` null means the conversation is still choosing a document; the
response's `selectedDocument` appears only on the turn where the assistant
picks (or switches) one. `updates` is a sparse {field name: value} patch —
the frontend owns the canonical fields state and merges it.
"""

from pydantic import Field

from app.schemas.camel import CamelModel
from app.schemas.chat import ChatMessage
from app.schemas.nda import ISO_DATE_RE


class DocChatRequest(CamelModel):
    transcript: list[ChatMessage]
    document_key: str | None = None
    fields: dict[str, str] = Field(default_factory=dict)
    today: str = Field(pattern=ISO_DATE_RE)


class DocChatResponse(CamelModel):
    reply: str
    selected_document: str | None = None
    updates: dict[str, str] = Field(default_factory=dict)
